### PR TITLE
topo: Detect Neuron devices as accelerators in addition to GPUs

### PR DESCRIPTION
**Pull request by rajachan**
.. and a couple other minor cleanups as separate commits. I can separate them out into a separate PR if folks prefer.

Testing: neuron-bench sweeps run successfully. Kicking off a CI trigger here for P* testing. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/nccl_ofi_sendrecv.c [hunk 1]
```diff
@@ -1782,11 +1779,14 @@ static inline int create_send_comm(nccl_net_ofi_conn_handle_t *handle,
...
 		NCCL_OFI_WARN("Could not allocate NCCL OFI requests free list for dev %d",
 			      device->base.dev_id);
-		free(ret_s_comm);
-		return ret;
+		goto out;
 	}
 
 	*s_comm = ret_s_comm;
+out:
+	if (ret)
```
Comment (rauteric) [inline src/nccl_ofi_sendrecv.c:1787]: `if (ret_s_comm)`?

File: src/nccl_ofi_topo.c [hunk 2]
```diff
@@ -18,8 +18,13 @@
 #include "nccl_ofi_topo.h"
 #include "nccl_ofi_math.h"
 
+#if HAVE_CUDA
 static const uint8_t display_controller_class_id = 0x03;
 static const unsigned short nvidia_vendor_id = 0x10de;
+#else
```
Comment (rauteric) [inline src/nccl_ofi_topo.c:24]: `#elif HAVE_NEURON`, for clarity if nothing else? Same for other similar `#else` clauses.

**Review by rauteric — CHANGES_REQUESTED** [review]

File: src/nccl_ofi_sendrecv.c (hunk 1 above)
Comment (rajachan) [reply src/nccl_ofi_sendrecv.c:1787]: The happy path won't be happy if I free it always.  I only want to free it if fi_getname() or the freelist init failed (with non-zero ret).

Comment (rauteric) [reply src/nccl_ofi_sendrecv.c:1787]: Oops, you're right 😊

File: src/nccl_ofi_topo.c [hunk 3]
```diff
@@ -18,8 +18,13 @@
...
+static const uint8_t system_peripheral_class_id = 0x08;
```
Comment (bwbarrett) [inline src/nccl_ofi_topo.c:25]: I wonder if it wouldn't be better to just have these named `target_class_id` and `target_vendor_id` and then not having the compile time conditional in `is_accelerator_dev()`.

**Review by bwbarrett — DISMISSED** [review]
In the future, I'd probably prefer splitting the new code and code cleanups into different PRs.

File: src/nccl_ofi_topo.c (hunk 3 above)
Comment (rajachan) [reply src/nccl_ofi_topo.c:25]: Yes, that is better, will fix.

File: src/nccl_ofi_topo.c [hunk 4]
```diff
@@ -664,15 +683,15 @@ static int propagate_gpu_group_counts(hwloc_topology_t topo)
...
-		ret = is_nvidia_pci_dev(obj, &is_gpu);
+		ret = is_accelerator_dev(obj, &is_accel);
 		if (ret != 0) {
 			NCCL_OFI_WARN("Error while checking whether hwloc topology node is nvidia GPU");
 			return ncclInternalError;
 		}
 
-		if (is_gpu) {
+		if (is_accel) {
 			propagate_gpu_count(obj);
```
Comment (maxtmann) [inline src/nccl_ofi_topo.c:695]: Picky nit: `propagate_accel_count()`?

**Review by rauteric — APPROVED** [review]

File: src/nccl_ofi_topo.c [hunk 5]
```diff
@@ -156,8 +161,18 @@ static int is_nvidia_pci_dev(hwloc_obj_t obj, bool *res)
 	   the class code. */
 	class_code = obj->attr->pcidev.class_id >> 8;
 
-	class_match = display_controller_class_id == class_code;
-	vendor_match = obj->attr->pcidev.vendor_id == nvidia_vendor_id;
+	/*
```
Comment (bwbarrett) [inline src/nccl_ofi_topo.c:164]: i probably would have moved the comment, too.  But it works.

**Review by bwbarrett — APPROVED** [review]

**Pull request merged**